src/zir_sema.zig

```diff
@@ -844,7 +844,71 @@ fn analyzeInstErrorSet(mod: *Module, scope: *Scope, inst: *zir.Inst.ErrorSet) In
 }
 
 fn analyzeInstMergeErrorSets(mod: *Module, scope: *Scope, inst: *zir.Inst.BinOp) InnerError!*Inst {
-    return mod.fail(scope, inst.base.src, "TODO implement merge_error_sets", .{});
+    const rhs_fields = (try resolveType(mod, scope, inst.positionals.rhs)).getErrs();
+    const lhs_fields = (try resolveType(mod, scope, inst.positionals.lhs)).getErrs();
+    if (lhs_fields == .anyerror or rhs_fields == .anyerror)
+        return mod.constInst(scope, inst.base.src, .{
+            .ty = Type.initTag(.type),
+            .val = Value.initTag(.anyerror_type),
+        });
+    // The declarations arena will store the hashmap.
+    var new_decl_arena = std.heap.ArenaAllocator.init(mod.gpa);
+    errdefer new_decl_arena.deinit();
+
+    const payload = try scope.arena().create(Value.Payload.ErrorSet);
+    payload.* = .{
+        .fields = .{},
+        .decl = undefined, // populated below
+    };
+    try payload.fields.ensureCapacity(&new_decl_arena.allocator, @intCast(u32, switch (rhs_fields) {
+        .err_single => 1,
+        .multiple => |mul| mul.size,
+        else => unreachable,
+    } + switch (rhs_fields) {
+        .err_single => 1,
+        .multiple => |mul| mul.size,
+        else => unreachable,
+    })); // TODO should we do this? only true when no overlapping
+
+    switch (lhs_fields) {
+        .err_single => |name| {
+            const entry = try mod.getErrorValue(name);
```

Comment:
This could just be `mod.global_error_set.get(name).?` or `getErrs()` for `.err_single` could just also return the value.